### ISSUE TITLE
support multithreading in HTTP server

### ISSUE THREAD

**Issue by dashmug** [golastmile/rasa_nlu#133]
I've looked at [rasa_nlu's HTTP API docs](http://rasa-nlu.readthedocs.io/en/latest/http.html) and it looks to me like rasa_nlu's HTTP server (`python -m rasa_nlu.server -c config.json`) is meant for development and testing and not for production as it cannot handle too many concurrent requests. Is this correct?

If so, what are your recommendations for running this in production? Do I need to write a proper wsgi app (like `flask` or `django`) that will use `rasa_nlu` as a library?

Thank you.

**Comment by amn41** [comment]
Well it's probably the multithreading that we need.  I know for example flask doesn't do that out of the box, you have to include something like gevent. http://flask.pocoo.org/docs/0.12/deploying/ So we could probably do something similar.

**Comment by dashmug** [comment]
For now, what I'm doing is to use a minimal `flask` + `gunicorn` wrapper that calls `rasa_nlu` like what is mentioned in [Using rasa NLU from Python](http://rasa-nlu.readthedocs.io/en/latest/python.html).

**Comment by jasonscharf** [comment]
Hey @dashmug, I run into issues where the HTTP server hangs, and quite frequently. This happens on the same inputs, so I don't think it's due to my models.

I run multiple instances of Rasa in containers to work around this. I'm not in production, but it's a simple enough solution. My Rasa endpoints respond quite fast, so it's fairly easy to see a threshold of sorts when they seem to be timing out.

There's a Dockerfile in the root. If you're not training in parallel, you could consider mounting read-only volumes w/ your models and serving via a bunch of `rasa_nlu.server` instances.

**Comment by jasonscharf** [comment]
Also, this is an amazing project. Thanks @amn41 and anyone else involved!

**Comment by dashmug** [comment]
@jasonscharf I don't have that problem. My Flask wrapper works as fast as the `rasa_nlu.server` so far.

**Comment by jasonscharf** [comment]
@amn41, cool. Good to know. I'm still trying to figure out why my endpoints hang, but isolating them in the short term works for me.

**Comment by amn41** [comment]
FYI to people watching this thread this is high prio. Ideally we would just wrap the current HTTPServer in a WSGIServer from gevent or something like that, rather than adding an additional dependency to flask. 

@jasonscharf your comment re running multiple instances - are they serving the same model or separate ones?

**Comment by dashmug** [comment]
That's good to hear @amn41!

**Comment by jasonscharf** [comment]
@amn41 multiple instances, although I'll double check - it's possible when I ran into the issue I describe I had multiple instances bound to the same volume and thus model. I can file a separate issue once I repro and get some more info.
